File: matter/player.js
const Matter = require('matter-js');
const CONFIG = require('./config.js');

class Player {
    constructor(world, engine, io) {
        this.world = world;
        this.engine = engine; 
        this.body = null;
        this.spawnX = 200;
        this.spawnY = 200;
        this.moveType = CONFIG.player.movement.type;
        this.sequence = 0;
        this.lastProcessedInput = null;
        this.io = io; 
        this.goals = 0; 
        this.assists = 0;
        this.points = 0;  
        this.side = 0; 
        // Ball kick config 
        const { normalKickDistance, normalKickVelocityAdd } = CONFIG.player.ballConfig;
        this.normalKickDistance = normalKickDistance;
        this.normalKickVelocityAdd = normalKickVelocityAdd;
    }
    afterPhysicsUpdate() {
        if (this.body) {
            const damping = CONFIG.player.physics.damping || 0.98;
            const xx = this.body.velocity.x * damping;
            const yy = this.body.velocity.y * damping;
            this.setVelocity(xx, yy);
        }
    }
    create(x = 200, y = 200) {
        this.spawnX = x;
        this.spawnY = y;
        this.body = this.createPhysicsBody(x, y);
        Matter.World.add(this.world, this.body);
        Matter.Events.on(this.engine, 'beforeUpdate', this.afterPhysicsUpdate.bind(this));
        return this;
    }

    createPhysicsBody(x, y) {
        const { radius } = CONFIG.player.graphic;
        const physics = CONFIG.player.physics;
        const categories = {
            outer: 0x0001,  // 00001
            inner: 0x0002,  // 00010
            player: 0x0004, // 00100
            ball: 0x0008,   // 01000
            net: 0x00016    // 10000
        };
        return Matter.Bodies.circle(x, y, radius, {
            label: 'player',
            mass: physics.mass,
            restitution: physics.restitution,
            friction: physics.friction,
            frictionAir: physics.frictionAir,
            frictionStatic: physics.frictionStatic,
            inertia: physics.inertia,
            slop: physics.slop,
            isStatic: false, 
            collisionFilter: {
                category: categories.player,
                mask: ~categories.inner  
            }
        });
    }

    update(input, ball) {
        if (this.moveType === 'force') {
            this.updateForceMovement(input);
        } else {
            this.updateVelocityMovement(input);
        }

        if (input.kick) {
            this.handleKick(ball);
        }

        this.lastProcessedInput = input.sequence;
    }

    updateForceMovement(input) {
        const { force, maxSpeed, deceleration } = CONFIG.player.movement;
        const velocity = this.body.velocity;
        const speed = Math.sqrt(velocity.x * velocity.x + velocity.y * velocity.y);

        if (speed < maxSpeed) {
            let fx = 0;
            let fy = 0;

            fx = input.inputX * force;
            fy = input.inputY * force;

            if (fx !== 0 && fy !== 0) {
                const normalize = 1 / Math.sqrt(2);
                fx *= normalize;
                fy *= normalize;
            }

            Matter.Body.applyForce(this.body, this.body.position, { x: fx, y: fy });
        }

        if (input.inputX === 0 && input.inputY === 0) {
            Matter.Body.setVelocity(this.body, {
                x: velocity.x * deceleration,
                y: velocity.y * deceleration
            });
        }
    }

    updateVelocityMovement(input) {
        const { maxSpeed, acceleration, deceleration } = CONFIG.player.movement;
        let vx = this.body.velocity.x;
        let vy = this.body.velocity.y;
        
        const deltaTime = input.deltaTime;
        if (input.inputX !== 0 || input.inputY !== 0) {
            // v = v0 + a*t
            vx += input.inputX * acceleration * deltaTime;
            vy += input.inputY * acceleration * deltaTime;
    
            // Limit max speed
            const currentSpeed = Math.sqrt(vx * vx + vy * vy);
            if (currentSpeed > maxSpeed) {
                const scale = maxSpeed / currentSpeed;
                vx *= scale;
                vy *= scale;
            }
        }
    
        Matter.Body.setVelocity(this.body, { x: vx, y: vy });
    }

    handleKick(ball) {
        const { player} = CONFIG;
        const distance = this.getDistanceTo(ball.body.position);
        const minDistance = (player.graphic.radius + ball.radius + this.normalKickDistance);
        if (distance <= minDistance) {
            const angle = this.getAngleTo(ball.body.position);

            const velocity = {
                x: Math.cos(angle) * this.normalKickVelocityAdd,
                y: Math.sin(angle) * this.normalKickVelocityAdd
            };

            const currentVelocity = ball.body.velocity;
            Matter.Body.setVelocity(ball.body, {
                x: currentVelocity.x + velocity.x,
                y: currentVelocity.y + velocity.y
            });
            this.io.emit('sendBallState', {
                position: {
                    x: ball.body.position.x,
                    y: ball.body.position.y
                },
                velocity: {
                    x: ball.body.velocity.x,
                    y: ball.body.velocity.y
                }, 
                timestamp: Date.now()
            });
            this.io.emit('kickingSound'); 
        }
    }

    getDistanceTo(position) {
        return Math.sqrt(
            Math.pow(this.body.position.x - position.x, 2) +
            Math.pow(this.body.position.y - position.y, 2)
        );
    }

    getAngleTo(position) {
        return Math.atan2(
            position.y - this.body.position.y,
            position.x - this.body.position.x
        );
    }

    reset() {
        Matter.Body.setVelocity(this.body, { x: 0, y: 0 });
        Matter.Body.setPosition(this.body, { x: this.spawnX, y: this.spawnY });
    }

    setVelocity(x, y) {
        Matter.Body.setVelocity(this.body, { x, y });
    }

    getVelocity() {
        return this.body.velocity;
    }

    setPosition(x, y) {
        Matter.Body.setPosition(this.body, { x, y });
    }

    getPosition() {
        return this.body.position;
    }

    getLastProcessedInput() {
        return this.lastProcessedInput;
    }
    destroy() {
        if (this.world) {
            Matter.Events.off(this.world, 'beforeUpdate', this.afterPhysicsUpdate);
        }
        if (this.body) {
            Matter.World.remove(this.world, this.body);
        }
    }
}

module.exports = Player;

File: player.py
import pymunk
import math

class Player:
    def __init__(self, space, config, player_id, gamePhysics, x=200, y=200):
        self.gamePhysics = gamePhysics 
        self.space = space
        self.last_processed_input = 0
        self.config = config
        self.player_id = player_id
        self.spawn_x = x
        self.spawn_y = y
        self.isFreezeNormalShooting = False
        self.freezeNormalShootingTimer = 0
        
        # Lấy config
        self.player_config = config["player"]
        self.physics_config = self.player_config["physics"]
        self.graphic_config = self.player_config["graphic"]
        self.movement_config = self.player_config["movement"]

        self.radius = self.graphic_config["radius"]
        # Khởi tạo physics body
        self.body, self.shape = self._create_physics_body()
        
        # Movement type
        self.movement_type = self.player_config["movement"]["type"]
        
        # Kick config
        self.normalKickDistance = self.player_config["ballConfig"]["normalKickDistance"]
        self.normalKickVelocityAdd = self.player_config["ballConfig"]["normalKickVelocityAdd"]
    
    def _create_physics_body(self):
        # Tạo body
        mass = self.physics_config["mass"]
        radius = self.graphic_config["radius"]
        moment = pymunk.moment_for_circle(mass, 0, radius)
        body = pymunk.Body(mass, moment)
        body.position = (self.spawn_x, self.spawn_y)

        # Tạo shape
        shape = pymunk.Circle(body, radius)
        shape.friction = self.physics_config["friction"]
        shape.elasticity = self.physics_config["restitution"]
        
        shape.filter = pymunk.ShapeFilter(categories = 0b1, mask = 0b11111011)
        # Các thuộc tính khác
        body.angular_damping = self.physics_config["frictionAir"]
        body.moment = self.physics_config["inertia"]
        
        # Collision type
        shape.collision_type = self.gamePhysics.playerCollisionType  # player collision type
        
        # Thêm vào space
        self.space.add(body, shape)
        
        return body, shape

    def solveGoThroughBall(self): 
        ball = self.gamePhysics.ball
        ball_pos = ball.get_position()
        player_pos = self.get_position()
        ball_radius = ball.radius
        player_radius = self.radius
        allow_through_distance = 1 
        
        dx = player_pos.x - ball_pos.x
        dy = player_pos.y - ball_pos.y
        distance = math.sqrt(dx*dx + dy*dy)

        min_distance = ball_radius + player_radius - allow_through_distance
    
        if distance < min_distance and distance > 0:  
            unit_x = dx / distance
            unit_y = dy / distance
            new_player_x = ball_pos.x + unit_x * (player_radius + ball_radius)
            new_player_y = ball_pos.y + unit_y * (player_radius + ball_radius)
            
            self.set_position(new_player_x, new_player_y)
            
            ball_vel = ball.get_velocity()
            player_vel = self.get_velocity()

    def update(self):
        damping = self.physics_config["damping"]
        current_velocity = self.body.velocity
        new_velocity = (
            current_velocity.x * damping,
            current_velocity.y * damping
        )
        self.body.velocity = new_velocity
        if self.isFreezeNormalShooting:
            self.freezeNormalShootingTimer -= 1/60  # Giảm timer mỗi frame (60 FPS)
            if self.freezeNormalShootingTimer <= 0:
                self.isFreezeNormalShooting = False
                self.freezeNormalShootingTimer = 0
    
    def apply_movement(self, movement_input):
        if movement_input["kick"] and not self.isFreezeNormalShooting:
            self.perform_kicking()
        
        if self.movement_type == "force":
            self._apply_force_movement(movement_input)
        else:
            self._apply_velocity_movement(movement_input)
        if movement_input["sequence"] > self.last_processed_input:
            self.last_processed_input = movement_input["sequence"]

    def _apply_force_movement(self, movement_input):
        force_magnitude = self.movement_config["force"]
        max_speed = self.movement_config["maxSpeed"]
        
        # Normalize diagonal movement
        if movement_input["x"] != 0 and movement_input["y"] != 0:
            normalize = 1 / math.sqrt(2)
            force_x = movement_input["x"] * force_magnitude * normalize
            force_y = movement_input["y"] * force_magnitude * normalize
        else:
            force_x = movement_input["x"] * force_magnitude
            force_y = movement_input["y"] * force_magnitude
            
        current_velocity = self.body.velocity
        current_speed = math.sqrt(current_velocity.x**2 + current_velocity.y**2)
        
        if current_speed < max_speed:
            self.body.apply_force_at_local_point((force_x, force_y), (0, 0))

    def _apply_velocity_movement(self, movement_input):
        # Config từ JS
        acceleration = self.movement_config["acceleration"]
        max_speed = self.movement_config["maxSpeed"] * 60  # Convert to units/second
        
        current_velocity = self.body.velocity
        delta_time = movement_input["deltaTime"]
        
        inputX = movement_input["inputX"]
        inputY = movement_input["inputY"]
        
        # Tính toán velocity mới với delta time từ client
        new_velocity_x = current_velocity.x + inputX * acceleration * 60 * delta_time
        new_velocity_y = current_velocity.y + inputY * acceleration * 60 * delta_time

        # Giới hạn max speed (units/second)
        current_speed = math.sqrt(new_velocity_x**2 + new_velocity_y**2)
        if current_speed > max_speed:
            scale = max_speed / current_speed
            new_velocity_x *= scale
            new_velocity_y *= scale
        
        self.body.velocity = (new_velocity_x, new_velocity_y)

    def perform_kicking(self):
        print("Performing kicking")  
        ball = self.gamePhysics.ball
        if ball:
            player_pos = self.body.position
            ball_pos = ball.body.position
            
            distance = math.sqrt((player_pos.x - ball_pos.x)**2 + (player_pos.y - ball_pos.y)**2)
            
            # Lấy tổng bán kính của player và ball
            total_radius = self.graphic_config["radius"] + ball.config["physics"]["radius"]
            print("Total Radius", total_radius)
            print("Normal Kick Distance: ", self.normalKickDistance)  
            print("Total Allowed Distance: ", total_radius + self.normalKickDistance)
            print("Distance between player and ball: ", distance)
            
            if distance <= self.normalKickDistance + total_radius:
                print("Shooting successfully") 
                # Tính vector hướng từ player đến ball
                direction_x = ball_pos.x - player_pos.x
                direction_y = ball_pos.y - player_pos.y
                
                # Normalize vector
                length = math.sqrt(direction_x**2 + direction_y**2)
                if length > 0:
                    direction_x /= length
                    direction_y /= length
                    
                # Add thêm velocity thay vì gán
                current_vel = ball.body.velocity
                ball.body.velocity = (
                    current_vel.x + direction_x * self.normalKickVelocityAdd * 60,
                    current_vel.y + direction_y * self.normalKickVelocityAdd * 60
                )
                self.isFreezeNormalShooting = True
                self.freezeNormalShootingTimer = self.player_config["ballConfig"]["normalKickFreezeTime"]

    def reset(self):
        self.body.position = (self.spawn_x, self.spawn_y)
        self.body.velocity = (0, 0)

    def get_position(self):
        return self.body.position

    def get_velocity(self):
        return self.body.velocity

    def set_velocity(self, x, y):
        self.body.velocity = (x, y)

    def set_position(self, x, y):
        self.body.position = (x, y)